profile ivalue for nvfuser

jjsjann123 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47668 profile ivalue for nvfuser**
* #47667 refactor profiling optional

Differential Revision: [D25255571](https://our.internmc.facebook.com/intern/diff/D25255571)

jjsjann123 committed:
profile ivalue for nvfuser

[ghstack-poisoned]

jjsjann123 committed:
Update on "profile ivalue for nvfuser"




[ghstack-poisoned]

jjsjann123 committed:
Update on "profile ivalue for nvfuser"




[ghstack-poisoned]

eellison commented:
@jjsjann123 can you get someone  from nvidia to review and accept before we land this ? @Krovatkin

tlemo commented on torch/csrc/jit/codegen/cuda/parser.cpp:
```diff
@@ -614,6 +617,79 @@ std::unordered_map<
+void profileIntList(ProfilingRecord* pr, Node* node, size_t offset) {
+  auto pn = insertProfileIValueOp(node, offset, pr);
+
+  std::function<void(Stack&)> ivalue_profiler = [pr, pn](Stack& stack) {
```
why assign to a std::function here? (rather than say ... `const auto`)

tlemo commented on torch/csrc/jit/codegen/cuda/parser.cpp:
```diff
@@ -614,6 +617,79 @@ std::unordered_map<
+void profileBool(ProfilingRecord* pr, Node* node, size_t offset) {
+  auto pn = insertProfileIValueOp(node, offset, pr);
+
+  std::function<void(Stack&)> ivalue_profiler = [pr, pn](Stack& stack) {
```
ditto

tlemo commented on torch/csrc/jit/codegen/cuda/parser.cpp:
```diff
@@ -638,6 +714,52 @@ bool isNodeParsible(const Node* node) {
   return IrParser::canParseNode(node);
 }
 
+// TODO: we should incorporate this to our parser as well;
```
It's not clear what our parser refers to in this context: replace `our parser` to a more explicit pointer ?

tlemo commented on torch/csrc/jit/codegen/cuda/parser.cpp:
```diff
@@ -638,6 +714,52 @@ bool isNodeParsible(const Node* node) {
+          ->schema();
+  if (node->matches(reduction_operator_schema)) {
+    switch (offset) {
+      case 1:
```
what do 1 and 2 mean here?

tlemo commented on torch/csrc/jit/codegen/cuda/parser.cpp:
```diff
@@ -638,6 +714,52 @@ bool isNodeParsible(const Node* node) {
+}
+
+void insertProfileNodesForCUDAFuser_(Block* block, ProfilingRecord* pr) {
+  for (auto it = block->nodes().begin(); it != block->nodes().end(); ++it) {
```
`for (const auto& n : block->nodes())` ?

tlemo commented on torch/csrc/jit/codegen/cuda/graph_fuser.cpp:
```diff
@@ -788,7 +855,9 @@ struct CudaGraphFuser {
 
     // TODO: failure in buildShapeExpressions should not break fusion execution,
     // we can add a try/catch here to bailout from removeOutputsUsedOnlyInSize.
+    GRAPH_DUMP("before build shape expression: ", graph_);
```
does this unconditionally dump the graph?

jjsjann123 replied:
Nope, this is only enabled via environment variable: https://github.com/pytorch/pytorch/blob/cbc34db39e6b4d2a9c685ba25416ed7e6df05890/torch/csrc/jit/jit_log.h#L6-L37

jjsjann123 committed:
Update on "profile ivalue for nvfuser"


Differential Revision: [D25255571](https://our.internmc.facebook.com/intern/diff/D25255571)

[ghstack-poisoned]

jjsjann123 committed:
Update on "profile ivalue for nvfuser"


Differential Revision: [D25255571](https://our.internmc.facebook.com/intern/diff/D25255571)

[ghstack-poisoned]

jjsjann123 committed:
Update on "profile ivalue for nvfuser"


Differential Revision: [D25255571](https://our.internmc.facebook.com/intern/diff/D25255571)

[ghstack-poisoned]

jjsjann123 committed:
Update on "profile ivalue for nvfuser"


Differential Revision: [D25255571](https://our.internmc.facebook.com/intern/diff/D25255571)

[ghstack-poisoned]

jjsjann123 committed:
Update on "profile ivalue for nvfuser"


Differential Revision: [D25255571](https://our.internmc.facebook.com/intern/diff/D25255571)

[ghstack-poisoned]

jjsjann123 committed:
Update on "profile ivalue for nvfuser"


Differential Revision: [D25255571](https://our.internmc.facebook.com/intern/diff/D25255571)

[ghstack-poisoned]

jjsjann123 committed:
Update on "profile ivalue for nvfuser"


Differential Revision: [D25255571](https://our.internmc.facebook.com/intern/diff/D25255571)

[ghstack-poisoned]

jjsjann123 committed:
Update on "profile ivalue for nvfuser"


Differential Revision: [D25255571](https://our.internmc.facebook.com/intern/diff/D25255571)

[ghstack-poisoned]

jjsjann123 committed:
Update on "profile ivalue for nvfuser"


Differential Revision: [D25255571](https://our.internmc.facebook.com/intern/diff/D25255571)

[ghstack-poisoned]

jjsjann123 committed:
Update on "profile ivalue for nvfuser"


Differential Revision: [D25255571](https://our.internmc.facebook.com/intern/diff/D25255571)

[ghstack-poisoned]

jjsjann123 committed:
Update on "profile ivalue for nvfuser"


Differential Revision: [D25255571](https://our.internmc.facebook.com/intern/diff/D25255571)

[ghstack-poisoned]

jjsjann123 committed:
Update on "profile ivalue for nvfuser"


Differential Revision: [D25255571](https://our.internmc.facebook.com/intern/diff/D25255571)

[ghstack-poisoned]
